plugin-server/package.json

```diff
@@ -21,6 +21,7 @@
         "prettier": "prettier --write .",
         "prettier:check": "prettier --check .",
         "prepublishOnly": "pnpm build",
+        "schema:build:json": "ts-json-schema-generator -f tsconfig.json --path 'src/ingestion-schema.ts' --no-type-check > src/schema.json && prettier --write src/schema.json",
```

Comment:
I’m a bit surprised that we generate the jsonschema from the type, as I am using to do it the other way: the schema is the source of truth, and the type is generated from it.

In your approach, I can change the SummarizedSessionRecordingEvent type, re-generate the jsonschema, and I’ll still pass validation while breaking CH? What failure mode is prevented by this schema then?

Reply:
I think this would make sense if the schema definition files here were placed into a shared location and the consumer used those definitions for their consumption. This works really well with Protobuf because you are unable to make backwards incompatible changes to a schema so there's little risk.

In this case you would have to share the schema with clickhouse in order to make this schema validation useful. With how things are setup currently this won't actually catch anything that would stop ingestion because if you update a field from `str`->`int` the json schema will be updated and the only chance you would be able to catch this is during code review.

If you do what @xvello is suggesting (and I think is the correct way to do it) you end up with a separate, usually manually configured, set of schema definitions and when someone inadvertently updates the schema the tests will catch the problem if they are setup correctly.

What you have here will work well _if_ you use this as the canonical schema definition that you will use to build replacement services with like you were using here. Let's say here Service 1 defines the schema and you use that to validate Service 2's outputs match what we expect. In the case of this outage we would have caught the issue that cropped up.

Reply:
On a totally separate topic I would make a folder for schemas since this will typically become huge and need to be separated out. _Especially_ if we move to something like protobufs.

Might I suggest something that looks like:
https://github.com/PostHog/posthog/tree/9e2596540745a05943b34c8950be5264d581aceb/posthog/idl

Reply:
I was going to make the argument that I can change the schema generate the type and it be incorrect - since my suggestion would be that it is way nicer to write typescript types than json schema...

Then I added a test to prove this would have caught yesterday's issue where active duration was a float - and you can't because the TS type is `number` and so I'd have to change the interface to be

```
<snip/>
/**
* @asType integer
*/
active_duration: number
<snip/>
```

yuck...

Reply:
I think that's the sign that we overgrew json and should look into structured binary format such as captnproto and protobuf. The validation overhead will be regained by faster marshalling/unmarshalling.

Looks like [clickhouse can read both formats](https://clickhouse.com/docs/en/interfaces/formats#capnproto), kafka-ui can [decode proto](https://github.com/provectus/kafka-ui/blob/master/kafka-ui-api/src/main/java/com/provectus/kafka/ui/serdes/builtin/ProtobufFileSerde.java) not not captnproto.

Reply:
This change to the plugin server seems to just copy what we're doing in the main PostHog repo for the schema.... which in turn was just a "got it working this way, moving on" level implementation during the early days of project "data exploration".

We've been talking about flipping the source of truth from TS types to Python dataclasses (nobody wants to hand-crank JSON schemas), but this has not been important enough for us to make progress.

Reply:
> This change to the plugin server seems to just copy what we're doing in the main PostHog repo for the schema

Yep, was thinking "if it ain't broke don't fix it" 😊 

> from TS types to Python dataclasses

which isn't possible for plugin server 🙃 

> (nobody wants to hand-crank JSON schemas),

#preach :)